matrix_content_scanner/scanner/scanner.py

```diff
@@ -59,14 +93,140 @@ async def scan_file(
+                return media
+
+            logger.warning(
+                "Media has changed since last scan, discarding cached result and"
+                " scanning again"
+            )
+
+            del self._result_cache[cache_key]
+
+        # Download and scan the file.
+        try:
+            media, cacheable = await self._download_and_scan(
+                media_path, metadata, thumbnail_params, media
+            )
+        except FileDirtyError as e:
+            if e.cacheable:
+                logger.info("Caching scan failure")
+
+                # If the test fails, don't store the media to save memory.
+                self._result_cache[cache_key] = CacheEntry(
+                    result=False,
+                    media=None,
+                    info=e.info,
+                )
+
+            raise
+
+        # Update the cache if the result should be cached.
+        if cacheable:
+            logger.info("Caching scan success")
+
+            cached_media: Optional[MediaDescription] = media
+
+            if (
+                self._max_size_to_cache is not None
+                and len(media.content) > self._max_size_to_cache
+            ):
+
+                # Don't cache the file's content if it exceeds the maximum allowed file
+                # size, to minimise memory usage.
+                logger.info(
+                    "File content has size %s, which is more than %s, not caching content",
+                    format_size(len(media.content)),
+                    format_size(self._max_size_to_cache),
+                )
+
+                cached_media = None
+
+            # Hash the media, that way if we need to re-download the file we can make sure
+            # it's the right one.
+            media_hash = hashlib.sha256(media.content).digest()
+
+            self._result_cache[cache_key] = CacheEntry(
+                result=True,
+                media=cached_media,
+                media_hash=media_hash,
+            )
+
+        return media
+
+    async def _download_and_scan(
+        self,
+        media_path: str,
+        metadata: Optional[JsonDict] = None,
+        thumbnail_params: Optional[Dict[str, List[str]]] = None,
+        media: Optional[MediaDescription] = None,
+    ) -> Tuple[MediaDescription, bool]:
+        """Downloads and scans the given file.
 
         Args:
             media_path: The `server_name/media_id` path for the media.
             metadata: The metadata attached to the file (e.g. decryption key), or None if
                 the file isn't encrypted.
             thumbnail_params: If present, then we want to request and scan a thumbnail
                 generated with the provided parameters instead of the full media.
+            media: The already downloaded media. If provided, the download step is
+                skipped. Usually provided if we've re-downloaded a file with a cached
+                result, but the file changed since the initial scan.
```

Comment:
the fact this is optional makes this function's API rather unclean. Maybe it should be called `_download_if_we_dont_already_have_it_and_scan`.

Or: why not move the call to `download_file` into the caller?

Reply:
> Or: why not move the call to download_file into the caller?

That's probably much better indeed, thanks. I wasn't really happy with the shape of the API but failed to think of something better.